### PR TITLE
Feat/spam

### PR DESCRIPTION
**Pull request by shadow2kk**

### CODE REVIEW

**Review by LucasM19 — APPROVED** [review]

File: apps/app/components/Inbox/index.tsx [hunk 1]
```diff
@@ -189,6 +190,33 @@ export const InboxComponent: React.FC = () => {
...
+              const newIds =
+                Object.keys(chooseMap).filter((key) => chooseMap[key]) ?? []
+              const seenIds = refSeenBoxList?.current?.getChooseIds() ?? []
+              const ids = [...newIds, ...seenIds]
+
+              if (!ids.length) return
+              try {
+                await api.batchMoveMessage(ids, Mailboxes.Spam)
+                if (newIds.length) {
+                  const map: Record<string, boolean> = {}
```
Comment (Codemasker) [inline apps/app/components/Inbox/index.tsx:203]:
```suggestion
                  const map = newIds.reduce<Record<string, boolean>>((acc, key) => ({ ...acc, [key]: true }), {})
```

File: apps/app/components/InfiniteMailbox/index.tsx [hunk 2]
```diff
@@ -109,12 +113,15 @@ const InfiniteBox: ForwardRefRenderFunction<
 
   const loaderEl = useMemo(() => loader || <Box>Loading</Box>, [loader])
 
-  const dataMsg: MessageItem[] = useMemo(() => {
+  const dataOriginFlat = useMemo(() => {
     if (!data) return []
-    const dataList = data.pages.map((item) => item.messages).flat()
+    return data.pages.map((item) => item.messages).flat()
```
Comment (Codemasker) [inline apps/app/components/InfiniteMailbox/index.tsx:118]:
```suggestion
    return data?.pages.map((item) => item.messages).flat() || []
```
In this way, `if` can be omitted

File: apps/app/components/InfiniteMailbox/index.tsx [hunk 3]
```diff
@@ -154,6 +161,17 @@ const InfiniteBox: ForwardRefRenderFunction<
   }, [chooseMap, parentChooseMap])
 
   useImperativeHandle(forwardedRef, () => ({
+    getChooseMsgs() {
+      return Object.keys(chooseMap)
+        .filter((key) => chooseMap[key])
+        .map((id) => {
+          let ret = null
+          dataOriginFlat.forEach((_item) => {
```
Comment (Codemasker) [inline apps/app/components/InfiniteMailbox/index.tsx:169]: You can use `Map<string, Item>`, `id` to `item`, reduce time complexity.

File: apps/app/components/Spam/index.tsx [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+  const dialog = useDialog()
+
+  const queryFn = useCallback(
+    async ({ pageParam = 0 }) => {
+      const { data } = await api.getMailboxesMessages(Mailboxes.Spam, pageParam)
+      // test empty status
+      // data.messages = []
+      return data
+    },
+    [api]
```
Comment (ghcoldsnap) [inline apps/app/components/Spam/index.tsx:44]: api 不需要加进去

**Review by ghcoldsnap — APPROVED** [review]

File: apps/app/components/Preview/index.tsx [hunk 5]
```diff
@@ -83,6 +95,7 @@ const PreviewContent = styled(Box)`
 export const PreviewComponent: React.FC = () => {
   const [t] = useTranslation('mailboxes')
   const [t2] = useTranslation('common')
+  const [t3] = useTranslation('preview')
```
Comment (Codemasker) [inline apps/app/components/Preview/index.tsx:98]:
```suggestion
  const [previewText] = useTranslation('preview')
```

File: apps/app/components/Preview/index.tsx [hunk 6]
```diff
@@ -83,6 +95,7 @@ const PreviewContent = styled(Box)`
 export const PreviewComponent: React.FC = () => {
   const [t] = useTranslation('mailboxes')
   const [t2] = useTranslation('common')
```
Comment (Codemasker) [inline apps/app/components/Preview/index.tsx:97]:
```suggestion
  const [commonText] = useTranslation('common')
```

File: apps/app/components/Preview/index.tsx [hunk 7]
```diff
@@ -83,6 +95,7 @@ const PreviewContent = styled(Box)`
 export const PreviewComponent: React.FC = () => {
   const [t] = useTranslation('mailboxes')
```
Comment (Codemasker) [inline apps/app/components/Preview/index.tsx:96]:
```suggestion
  const [mailboxesText] = useTranslation('mailboxes')
```

File: apps/app/components/Spam/index.tsx [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+  const [isChooseMode, setIsChooseMode] = useState(false)
+  const userProps = useAtomValue(userPropertiesAtom)
+
+  const toast = useToast()
+  const dialog = useDialog()
+
+  const queryFn = useCallback(
+    async ({ pageParam = 0 }) => {
+      const { data } = await api.getMailboxesMessages(Mailboxes.Spam, pageParam)
+      // test empty status
```
Comment (Codemasker) [inline apps/app/components/Spam/index.tsx:40]: The comment should be deleted

File: apps/app/components/Spam/index.tsx [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+              })
+            },
+            [BulkActionType.NotSpam]: async () => {
+              const msg = refBoxList?.current?.getChooseMsgs()
+              if (!msg?.length) return
+
+              const sendIds: string[] = []
+              const inboxIds: string[] = []
+
+              msg.forEach((item) => {
```
Comment (Codemasker) [inline apps/app/components/Spam/index.tsx:89]:
```suggestion
              const { sendIds, inboxIds } =  msg.reduce((acc, item) => {
                if (!item) return acc
                if (
                  userProps?.aliases.some(
                    (_item: { address: string }) =>
                      _item.address === item?.from.address
                  )
                ) {
                  return { ...acc, sendIds: [...acc.sendIds, item.id] }
                } else {
                  return { ...acc, inboxIds: [...acc.inboxIds, item.id] }
                }
              }, { sendIds: [], inboxIds: [] })
```

File: apps/app/components/Spam/index.tsx [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+            fontWeight="700"
+            fontSize="24px"
+            lineHeight="30px"
+          >
+            {t('spam.title')}
+          </WrapItem>
+        </Wrap>
+
+        <Spacer />
+        {/* <Button
```
Comment (Codemasker) [inline apps/app/components/Spam/index.tsx:134]: The comment should be deleted

File: apps/app/i18n/en.json [hunk 11]
```diff
@@ -152,6 +159,14 @@
                 "ok": "The message has been restore.",
                 "fail": "Restore is fail, try again."
             },
+            "spam": {
+                "ok": "The messages have been moved to spam",
+                "fail": "Move fail, try again."
+            },
+            "notSpam": {
```
Comment (Codemasker) [inline apps/app/i18n/en.json:166]: `Camel-Case` should not be used here

**Review by Codemasker — APPROVED** [review]

**Pull request merged**